Add support for AWS session tokens

DaveCTurner opened the pull request:
AWS supports the creation and use of credentials that are only valid for a
fixed period of time. These credentials comprise three parts: the usual access
key and secret key, together with a session token. This commit adds support for
these three-part credentials to the EC2 discovery plugin and the S3 repository
plugin.

Note that session tokens are only valid for a limited period of time and yet
there is no mechanism for refreshing or rotating them when they expire without
restarting Elasticsearch.  Nonetheless, this feature is already useful for
nodes that need only run for a few days, such as for training, testing or
evaluation. #29135 tracks the work towards allowing these credentials to be
refreshed at runtime.

Resolves #16428

elasticmachine commented:
Pinging @elastic/es-distributed

elasticmachine commented:
Pinging @elastic/es-security

tlrx commented on plugins/repository-s3/src/test/java/org/elasticsearch/repositories/s3/S3ClientSettingsTests.java:
```diff
@@ -0,0 +1,129 @@
+        assertThat(defaultSettings.proxyPassword, isEmptyString());
+        assertThat(defaultSettings.readTimeoutMillis, is(ClientConfiguration.DEFAULT_SOCKET_TIMEOUT));
+        assertThat(defaultSettings.maxRetries, is(ClientConfiguration.DEFAULT_RETRY_POLICY.getMaxErrorRetry()));
+        assertThat(defaultSettings.throttleRetries, is(ClientConfiguration.DEFAULT_THROTTLE_RETRIES));
```
Thanks for adding these

tlrx commented on plugins/repository-s3/src/test/java/org/elasticsearch/repositories/s3/S3ClientSettingsTests.java:
```diff
@@ -0,0 +1,129 @@
+            () -> {
+                final MockSecureSettings secureSettings = new MockSecureSettings();
+                secureSettings.setString("s3.client.default.access_key", "aws_key");
+                S3ClientSettings.load(Settings.builder().setSecureSettings(secureSettings).build());
```
Nit: you can build the settings outside the expectThrows block, I find it easier to "spot" where the exception is expected to be thrown

DaveCTurner replied:
++ I pushed ee41de633714c0797dd2823595d4a46afa6c68b6

tlrx commented on docs/plugins/discovery-ec2.asciidoc:
```diff
@@ -46,6 +46,11 @@ Those that must be stored in the keystore are marked as `Secure`.
 
+`session_token`::
+
+    An s3 session token. The `access_key` and `secret_key` settings must also
```
Nit: An -> A

tlrx replied:
Did I just correct an English about his English grammar? Yes I did. Was I right? Absolutly not.

Thanks :)

tlrx commented on docs/plugins/repository-s3.asciidoc:
```diff
@@ -73,6 +73,11 @@ are marked as `Secure`.
 
+`session_token`::
+
+    An s3 session token. The `access_key` and `secret_key` settings must also
```
Nit: An -> A

DaveCTurner commented:
@tlrx I duplicated the QA tests as we discussed, and tightened up the auth checks in `AmazonS3TestServer` a bit in order to check that the right bucket is being accessed. This is ready for another look.

tlrx commented on plugins/repository-s3/qa/amazon-s3/src/test/java/org/elasticsearch/repositories/s3/AmazonS3TestServer.java:
```diff
@@ -106,13 +106,38 @@ public Response handle(final String method,
         if (handler != null) {
+            final String bucket = params.get("bucket");
+            if (bucket != null && permittedBucket.equals(bucket) == false) {
+                // allow a null bucket to support bucket-free APIs like ListBuckets?
```
>  allow a null bucket to support bucket-free APIs like ListBuckets?

I don't think it's needed, returning an error seems the right thing to do

DaveCTurner replied:
It is needed, but it's not `ListBuckets` so I pushed 63caae2bd24. It's the multi-object delete API, which puts the bucket name in the `Host` header rather than in a URL parameter, which I do not think is worth adding the extra plumbing for.

tlrx commented on plugins/repository-s3/qa/amazon-s3/build.gradle:
```diff
@@ -39,16 +39,37 @@ forbiddenApisTest {
-    s3Bucket = 'bucket_test'
-    s3BasePath = 'integration_test'
+// We test against two repositories, one which uses the usual two-part "permanent" credentials and
+// the other which uses three-part "temporary" or "session" credentials.
```
Hopefully we'll split this into two QA projects in a follow up PR, so for now I'd prefer to not rename the existing env vars. 

I understand the permanent/temporary renaming for repositories or buckets but I still think that the env vars names should reflect the feature implemented, ie keeping `amazon_s3_access_key` for the default use case, and maybe `amazon_s3_access_key_for_session_tokens`.

DaveCTurner replied:
I have reverted the changes to the names of the existing environment variables.

tlrx commented on plugins/repository-s3/qa/amazon-s3/build.gradle:
```diff
@@ -39,16 +39,37 @@ forbiddenApisTest {
+// If all these variables are missing then we are testing against the internal fixture instead, which has the following
+// credentials hard-coded in.
+
+if (!s3PermanentAccessKey && !s3PermanentSecretKey && !s3PermanentBucket && !s3PermanentBasePath
```
Can we set the defaults in separate condition blocks? So that if a set [amazon_s3_access_key/amazon_s3_secret_key/amazon_s3_base_path/amazon_s3_bucket] is set it runs against the real service whatever is set for `amazon_s3_access_key_for_session_tokens`

DaveCTurner replied:
As it currently stands, I think it makes more sense to either test entirely against the fake fixture or entirely against the real service (and therefore not run the fake fixture). If there is a followup that allows these two cases to run separately then hopefully it's not too hard to disentangle this then.

tlrx commented on plugins/repository-s3/qa/amazon-s3/build.gradle:
```diff
@@ -57,26 +78,34 @@ task s3Fixture(type: AntFixture) {
     env 'CLASSPATH', "${ -> project.sourceSets.test.runtimeClasspath.asPath }"
     executable = new File(project.runtimeJavaHome, 'bin/java')
-    args 'org.elasticsearch.repositories.s3.AmazonS3Fixture', baseDir, s3Bucket
+    args 'org.elasticsearch.repositories.s3.AmazonS3Fixture', baseDir
```
I'd prefer if we continue to pass the bucket names to the fixture so that we can reuse it in different projects more easily.

DaveCTurner replied:
Ok, I've reverted that. There is, however, already some other coupling (duplication of string values) between the `build.gradle` file and the fake S3 fixture that would need attention to use this fixture elsewhere.

tlrx commented on plugins/repository-s3/qa/amazon-s3/src/test/java/org/elasticsearch/repositories/s3/AmazonS3Fixture.java:
```diff
@@ -53,8 +53,8 @@
     public static void main(String[] args) throws Exception {
-        if (args == null || args.length != 2) {
-            throw new IllegalArgumentException("AmazonS3Fixture <working directory> <bucket>");
+        if (args == null || args.length != 1) {
```
Can we revert this please?

DaveCTurner replied:
We now pass the bucket names in on the command line again.

tlrx commented on plugins/repository-s3/qa/amazon-s3/src/test/java/org/elasticsearch/repositories/s3/AmazonS3Fixture.java:
```diff
@@ -72,7 +72,8 @@ public static void main(String[] args) throws Exception {
             final AmazonS3TestServer storageTestServer = new AmazonS3TestServer(storageUrl);
-            storageTestServer.createBucket(args[1]);
+            storageTestServer.createBucket("permanent_bucket_test");
+            storageTestServer.createBucket("temporary_bucket_test");
```
Can we revert this please, and pass the bucket names as program args?

DaveCTurner replied:
We now pass the bucket names in on the command line again.

DaveCTurner commented:
@tlrx I have merged master, which was nontrivial thanks to #31383, and addressed some of your comments. I can't see how these two cases (permanent vs temporary credentials) are eventually going to work, so I've held off on any changes that depends on that. I would prefer to merge this and proceed from there, but if you think anything here will prove obstructive then it'd be better not to.
